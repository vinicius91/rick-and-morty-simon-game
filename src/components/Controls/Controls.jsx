// @flow

import React, { Component } from 'react';
import {
  ScoreWrapper,
  Score,
  Label,
  Wrapper,
  ToggleWrapper,
  SwitchWrapper,
  Switch,
  BtnWrapper,
  StartBtn,
  BtnLabel,
  ActionsWrapper
} from './Controls.styles';
import type { Ids } from '../../types/character';
import type { ControlsProps } from '../../types/controls';

class ControlsComponent extends Component<ControlsProps> {
  audioRef: ?HTMLAudioElement;

  componentDidUpdate() {
    const {
      isMachineStep,
      isUserStep,
      score,
      ids,
      sequence,
      nextStageGame,
      isError
    } = this.props;
    if (isMachineStep && !isUserStep) {
      nextStageGame(score, ids, sequence);
    }
    if (isError) {
      if (this.audioRef && this.audioRef.play) {
        this.audioRef.play();
      }
    }
  }

  handleStartClick = (ids: Ids) => {
    const { isOn, initializeGame } = this.props;
    if (isOn) {
      initializeGame(ids);
    }
  };

  render() {
    const errorSound = 'https://simongame.s3.amazonaws.com/errorSound.mp3';
    const { score, isActive, isOn, ids, loading, toggleGame } = this.props;
    if (!loading) {
      return (
        <Wrapper>
          <audio
            ref={(audio: ?HTMLAudioElement) => {
              this.audioRef = audio;
            }}
            src={errorSound}
          />
          <ScoreWrapper>
            <Score active={isOn}>{score ? score.toFixed(0) : '--'}</Score>
            <Label>COUNT</Label>
          </ScoreWrapper>
          <ActionsWrapper>
            <BtnWrapper>
              <StartBtn
                active={isActive}
                enabled={isOn}
                onClick={() => this.handleStartClick(ids)}
              />
              <BtnLabel>START</BtnLabel>
            </BtnWrapper>
            <ToggleWrapper>
              <Label>OFF</Label>
              <SwitchWrapper onClick={toggleGame}>
                <Switch active={isOn} />
              </SwitchWrapper>
              <Label>ON</Label>
            </ToggleWrapper>
          </ActionsWrapper>
        </Wrapper>
      );
    }
    return <div />;
  }
}

export default ControlsComponent;
