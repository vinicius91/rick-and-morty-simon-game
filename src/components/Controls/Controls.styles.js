import styled, { keyframes, css } from 'styled-components';

const wrapperBackground = css`
  background: #e2e2e2;
  background: -moz-linear-gradient(
    -45deg,
    #e2e2e2 0%,
    #dbdbdb 50%,
    #d1d1d1 51%,
    #fefefe 100%
  );
  background: -webkit-linear-gradient(
    -45deg,
    #e2e2e2 0%,
    #dbdbdb 50%,
    #d1d1d1 51%,
    #fefefe 100%
  );
  background: linear-gradient(
    135deg,
    #e2e2e2 0%,
    #dbdbdb 50%,
    #d1d1d1 51%,
    #fefefe 100%
  );
  filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#e2e2e2', endColorstr='#fefefe',GradientType=1 );
`;

export const Wrapper = styled.div`
  ${wrapperBackground}
  margin: 10px;
  min-height: 60px;
  border-radius: 10px;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-content: center;
  width: 100%;
  padding: 10px;
  box-shadow: 0 4px 5px -2px rgba(0, 0, 0, 0.2),
    0 7px 10px 1px rgba(0, 0, 0, 0.14), 0 2px 16px 1px rgba(0, 0, 0, 0.12);
`;

export const ScoreWrapper = styled.div`
  @media (max-width: 480px) {
    margin: 0px 15px;
  }
  flex-grow: 0;
  text-align: center;
`;

export const Score = styled.h1`
  color: ${props => (props.active ? '#dc0d29' : '#222')};
  transition: color 0.2s ease-in 0.1s;
  padding: 0px;
  font-size: 3em;
  height: 60px;
  width: 60px;
  background-color: #32050c;
  position: relative;
  border: 4px solid #222;
  border-radius: 10px;
  margin: auto;
`;

export const Label = styled.h3`
  @media (max-width: 400px) {
    font-size: 0.7em;
  }
  display: inline-block;
  color: #222;
  font-size: 1em;
  text-align: center;
  margin: 5px;
`;

export const ActionsWrapper = styled.div`
  flex-grow: 0;
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  margin-left: 10px;
`;

export const ToggleWrapper = styled.div`
  align-self: stretch;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  border: 0;
`;

export const SwitchWrapper = styled.div`
  width: 60px;
  height: 30px;
  background-color: #32050c;
  border-radius: 2px;
  cursor: pointer;
`;

const toggleColor = keyframes`
  from {
    background-color: #dc5245;
  }
  to {
    background-color: #5fd382;
  }
`;

const toggleAnimation = css`
  transform: translateX(29px);
  animation: ${toggleColor} 1s both;
`;

export const Switch = styled.div`
  position: relative;
  width: 26px;
  height: 26px;
  background-color: #dc5245;
  border-radius: 4px;
  border: 2px solid #333;
  transition: transform 0.2s ease-in 0.1s;

  ${props => (props.active ? toggleAnimation : '')}
`;

export const BtnWrapper = styled.div`
  flex-grow: 0;
  margin: 5px 10px;
  display: inline-block;
`;

const activeButtonMixin = css`
  cursor: pointer;

  :active {
    width: 19px;
    height: 19px;
    margin-bottom: 1px;
  }
`;

export const ActionBtn = styled.div`
  width: 20px;
  height: 20px;
  position: relative;
  border-radius: 100%;
  border: 4px solid #333;
  box-shadow: 0px 2px 3px #222;
  margin: auto;
  top: -5px;

  ${props => (props.enabled ? activeButtonMixin : '')}
`;

export const BtnLabel = styled(Label)`
  display: block;
  margin: auto;
  font-size: 0.9em;
`;

export const StartBtn = styled(ActionBtn)`
  background-color: #9f0f17;
  ${props =>
    props.active
      ? ` box-shadow: 0 0 5px #fff, 
    inset 0 0 5px #fff, 
    0 0 10px #9f0f17,
    inset 0 0 10px #9f0f17;`
      : ''}
`;
