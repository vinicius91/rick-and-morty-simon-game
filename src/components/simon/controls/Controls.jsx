// @flow

import * as React from 'react';
import {
  ScoreWrapper,
  Score,
  Label,
  Wrapper,
  ToggleWrapper,
  SwitchWrapper,
  Switch,
  BtnWrapper,
  StartBtn,
  StrictBtn,
  BtnLabel,
  ActionsWrapper
} from './Controls.styles';

type Props = {
  score?: number,
  isStrict?: boolean,
  isActive?: boolean,
  isOn?: boolean,
  strictClickHandler: () => void,
  startClickHandler: () => void,
  gameToggleHandler: () => void
};
const Controls = (props: Props) => {
  const {
    score,
    isStrict,
    isActive,
    isOn,
    strictClickHandler,
    startClickHandler,
    gameToggleHandler
  } = props;
  return (
    <Wrapper>
      <ScoreWrapper>
        <Score active={isOn}>{score ? score.toFixed(0) : '--'}</Score>
        <Label>COUNT</Label>
      </ScoreWrapper>
      <ActionsWrapper>
        <BtnWrapper>
          <StartBtn active={isActive} onClick={startClickHandler} />
          <BtnLabel>START</BtnLabel>
        </BtnWrapper>
        <BtnWrapper>
          <StrictBtn active={isStrict} onClick={strictClickHandler} />
          <BtnLabel>STRICT</BtnLabel>
        </BtnWrapper>
        <ToggleWrapper>
          <Label>OFF</Label>
          <SwitchWrapper onClick={gameToggleHandler}>
            <Switch active={isOn} />
          </SwitchWrapper>
          <Label>ON</Label>
        </ToggleWrapper>
      </ActionsWrapper>
    </Wrapper>
  );
};

Controls.defaultProps = {
  score: 0,
  isStrict: false,
  isActive: false,
  isOn: false
};

export default Controls;
