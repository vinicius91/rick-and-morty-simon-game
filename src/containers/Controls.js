import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import {
  toggleGame,
  toggleGameMode,
  initializeGame,
  nextStageGame
} from '../store/actions/game';

import ControlsComponent from '../components/Controls/Controls';

import type { State } from '../types/index';

const mapStateToProps = (state: State) => {
  const {
    score,
    isStrict,
    isActive,
    isOn,
    ids,
    isError,
    sequence,
    isMachineStep,
    isUserStep
  } = state.game;

  return {
    score,
    isStrict,
    isActive,
    isOn,
    ids,
    isError,
    sequence,
    isMachineStep,
    isUserStep,
    loading: state.characters.loading
  };
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      toggleGame,
      toggleGameMode,
      initializeGame,
      nextStageGame
    },
    dispatch
  );

const ControlsContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ControlsComponent);

export default ControlsContainer;
