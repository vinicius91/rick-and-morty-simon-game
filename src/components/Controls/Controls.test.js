import React from 'react';
import renderer from 'react-test-renderer';
import Controls from './Controls';

describe('Controls Component', () => {
  const fixture = {
    score: 0,
    isActive: false,
    isOn: false,
    loading: false,
    isMachineStep: false,
    isUserStep: false,
    ids: [1, 2, 3, 4],
    isError: false,
    sequence: [1],
    toggleGame: () => {},
    initializeGame: () => {},
    nextStageGame: () => {}
  };

  it('Should render with default', () => {
    const component = renderer.create(<Controls {...fixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render toggle on', () => {
    const customFixture = { ...fixture, isOn: true };
    const component = renderer.create(<Controls {...customFixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render without game active', () => {
    const customFixture = { ...fixture, isOn: true, isActive: true };
    const component = renderer.create(<Controls {...customFixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render without game active and score', () => {
    const customFixture = { ...fixture, isOn: true, isActive: true, score: 10 };
    const component = renderer.create(<Controls {...customFixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render blank with loading', () => {
    const customFixture = { ...fixture, loading: true };
    const component = renderer.create(<Controls {...customFixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
