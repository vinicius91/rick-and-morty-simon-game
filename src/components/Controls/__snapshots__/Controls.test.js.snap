// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls Component Should render blank with loading 1`] = `<div />`;

exports[`Controls Component Should render toggle on 1`] = `
<div
  className="Controlsstyles__Wrapper-sc-1qy04f2-0 hvVbvp"
>
  <audio
    src="https://simongame.s3.amazonaws.com/errorSound.mp3"
  />
  <div
    className="Controlsstyles__ScoreWrapper-sc-1qy04f2-1 cPQxNE"
  >
    <h1
      className="Controlsstyles__Score-sc-1qy04f2-2 bLancD"
    >
      --
    </h1>
    <h3
      className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
    >
      COUNT
    </h3>
  </div>
  <div
    className="Controlsstyles__ActionsWrapper-sc-1qy04f2-4 cUmXVB"
  >
    <div
      className="Controlsstyles__BtnWrapper-sc-1qy04f2-8 hIyuSu"
    >
      <div
        className="Controlsstyles__ActionBtn-sc-1qy04f2-9 Controlsstyles__StartBtn-sc-1qy04f2-11 juknOM"
        onClick={[Function]}
      />
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 Controlsstyles__BtnLabel-sc-1qy04f2-10 iqMRns"
      >
        START
      </h3>
    </div>
    <div
      className="Controlsstyles__ToggleWrapper-sc-1qy04f2-5 jSBQeW"
    >
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        OFF
      </h3>
      <div
        className="Controlsstyles__SwitchWrapper-sc-1qy04f2-6 iJhJgk"
        onClick={[Function]}
      >
        <div
          className="Controlsstyles__Switch-sc-1qy04f2-7 bqrkBc"
        />
      </div>
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        ON
      </h3>
    </div>
  </div>
</div>
`;

exports[`Controls Component Should render with default 1`] = `
<div
  className="Controlsstyles__Wrapper-sc-1qy04f2-0 hvVbvp"
>
  <audio
    src="https://simongame.s3.amazonaws.com/errorSound.mp3"
  />
  <div
    className="Controlsstyles__ScoreWrapper-sc-1qy04f2-1 cPQxNE"
  >
    <h1
      className="Controlsstyles__Score-sc-1qy04f2-2 hTCxUG"
    >
      --
    </h1>
    <h3
      className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
    >
      COUNT
    </h3>
  </div>
  <div
    className="Controlsstyles__ActionsWrapper-sc-1qy04f2-4 cUmXVB"
  >
    <div
      className="Controlsstyles__BtnWrapper-sc-1qy04f2-8 hIyuSu"
    >
      <div
        className="Controlsstyles__ActionBtn-sc-1qy04f2-9 Controlsstyles__StartBtn-sc-1qy04f2-11 juknOM"
        onClick={[Function]}
      />
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 Controlsstyles__BtnLabel-sc-1qy04f2-10 iqMRns"
      >
        START
      </h3>
    </div>
    <div
      className="Controlsstyles__ToggleWrapper-sc-1qy04f2-5 jSBQeW"
    >
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        OFF
      </h3>
      <div
        className="Controlsstyles__SwitchWrapper-sc-1qy04f2-6 iJhJgk"
        onClick={[Function]}
      >
        <div
          className="Controlsstyles__Switch-sc-1qy04f2-7 jPJxzl"
        />
      </div>
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        ON
      </h3>
    </div>
  </div>
</div>
`;

exports[`Controls Component Should render without game active 1`] = `
<div
  className="Controlsstyles__Wrapper-sc-1qy04f2-0 hvVbvp"
>
  <audio
    src="https://simongame.s3.amazonaws.com/errorSound.mp3"
  />
  <div
    className="Controlsstyles__ScoreWrapper-sc-1qy04f2-1 cPQxNE"
  >
    <h1
      className="Controlsstyles__Score-sc-1qy04f2-2 bLancD"
    >
      --
    </h1>
    <h3
      className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
    >
      COUNT
    </h3>
  </div>
  <div
    className="Controlsstyles__ActionsWrapper-sc-1qy04f2-4 cUmXVB"
  >
    <div
      className="Controlsstyles__BtnWrapper-sc-1qy04f2-8 hIyuSu"
    >
      <div
        className="Controlsstyles__ActionBtn-sc-1qy04f2-9 Controlsstyles__StartBtn-sc-1qy04f2-11 kcvcuu"
        onClick={[Function]}
      />
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 Controlsstyles__BtnLabel-sc-1qy04f2-10 iqMRns"
      >
        START
      </h3>
    </div>
    <div
      className="Controlsstyles__ToggleWrapper-sc-1qy04f2-5 jSBQeW"
    >
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        OFF
      </h3>
      <div
        className="Controlsstyles__SwitchWrapper-sc-1qy04f2-6 iJhJgk"
        onClick={[Function]}
      >
        <div
          className="Controlsstyles__Switch-sc-1qy04f2-7 bqrkBc"
        />
      </div>
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        ON
      </h3>
    </div>
  </div>
</div>
`;

exports[`Controls Component Should render without game active and score 1`] = `
<div
  className="Controlsstyles__Wrapper-sc-1qy04f2-0 hvVbvp"
>
  <audio
    src="https://simongame.s3.amazonaws.com/errorSound.mp3"
  />
  <div
    className="Controlsstyles__ScoreWrapper-sc-1qy04f2-1 cPQxNE"
  >
    <h1
      className="Controlsstyles__Score-sc-1qy04f2-2 bLancD"
    >
      10
    </h1>
    <h3
      className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
    >
      COUNT
    </h3>
  </div>
  <div
    className="Controlsstyles__ActionsWrapper-sc-1qy04f2-4 cUmXVB"
  >
    <div
      className="Controlsstyles__BtnWrapper-sc-1qy04f2-8 hIyuSu"
    >
      <div
        className="Controlsstyles__ActionBtn-sc-1qy04f2-9 Controlsstyles__StartBtn-sc-1qy04f2-11 kcvcuu"
        onClick={[Function]}
      />
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 Controlsstyles__BtnLabel-sc-1qy04f2-10 iqMRns"
      >
        START
      </h3>
    </div>
    <div
      className="Controlsstyles__ToggleWrapper-sc-1qy04f2-5 jSBQeW"
    >
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        OFF
      </h3>
      <div
        className="Controlsstyles__SwitchWrapper-sc-1qy04f2-6 iJhJgk"
        onClick={[Function]}
      >
        <div
          className="Controlsstyles__Switch-sc-1qy04f2-7 bqrkBc"
        />
      </div>
      <h3
        className="Controlsstyles__Label-sc-1qy04f2-3 dUBbjf"
      >
        ON
      </h3>
    </div>
  </div>
</div>
`;
